File: python/features/steps/standup_window_steps.py
"""Behave steps for standup time window scenarios."""

from __future__ import annotations

import json
import os
import subprocess
import urllib.error
import urllib.request
from datetime import datetime, timedelta, timezone
from pathlib import Path
from typing import Optional
from zoneinfo import ZoneInfo

import yaml
from behave import given, then, when

from kanbus.config import DEFAULT_CONFIGURATION
from kanbus.standup import load_standup_configuration, resolve_standup_profile
from kanbus.standup_window import (
    StandupWindowOverrides,
    parse_standup_lookback_hours,
    resolve_standup_window_settings,
)

from features.steps.shared import (
    load_project_directory,
    read_issue_file,
    write_issue_file,
)


def _load_config_payload(context: object) -> dict:
    repository = Path(context.working_directory)
    config_path = repository / ".kanbus.yml"
    payload = yaml.safe_load(config_path.read_text(encoding="utf-8"))
    if not isinstance(payload, dict):
        payload = dict(DEFAULT_CONFIGURATION)
    return payload


def _write_config_payload(context: object, payload: dict) -> None:
    repository = Path(context.working_directory)
    config_path = repository / ".kanbus.yml"
    config_path.write_text(yaml.safe_dump(payload, sort_keys=False), encoding="utf-8")


def _standup_timezone(context: object) -> ZoneInfo:
    timezone_name = getattr(context, "standup_timezone_name", None)
    if timezone_name:
        return ZoneInfo(timezone_name)
    configuration = load_standup_configuration(Path(context.working_directory))
    settings = resolve_standup_window_settings(
        configuration,
        MEETING_SCRIPT_PROFILE,
        None,
    )
    return settings.timezone


MEETING_SCRIPT_PROFILE = "meeting-script"


def _report_time(context: object) -> datetime:
    override = getattr(context, "standup_report_time", None)
    if override is not None:
        return override
    return datetime.now(timezone.utc)


def _clear_report_time() -> None:
    os.environ.pop("KANBUS_STANDUP_REPORT_TIME", None)


def _set_report_time(context: object, report_time: datetime) -> None:
    context.standup_report_time = report_time
    os.environ["KANBUS_STANDUP_REPORT_TIME"] = report_time.isoformat(
        timespec="seconds"
    ).replace("+00:00", "Z")


def _ensure_live_report_time(context: object) -> None:
    if getattr(context, "standup_report_time", None) is None:
        _clear_report_time()


def _resolved_window_settings(
    context: object,
    profile: Optional[str] = None,
    apply_profile_defaults: bool = True,
) -> object:
    root = Path(context.working_directory)
    configuration = load_standup_configuration(root)
    resolved_profile = None
    if apply_profile_defaults:
        resolved_profile = resolve_standup_profile(
            profile or getattr(context, "standup_profile", None)
        )
    overrides = getattr(context, "standup_window_overrides", StandupWindowOverrides())
    return resolve_standup_window_settings(configuration, resolved_profile, overrides)


@given('standup window is "{window}"')
def given_standup_window(context: object, window: str) -> None:
    """Set standup window in project configuration."""
    _ensure_live_report_time(context)
    payload = _load_config_payload(context)
    payload.setdefault("standup", {})
    payload["standup"]["window"] = window
    _write_config_payload(context, payload)


@given('standup lookback is "{lookback}"')
def given_standup_lookback(context: object, lookback: str) -> None:
    """Set standup lookback in project configuration."""
    payload = _load_config_payload(context)
    payload.setdefault("standup", {})
    payload["standup"]["lookback"] = lookback
    _write_config_payload(context, payload)


@given("standup skip_weekends is {enabled}")
def given_standup_skip_weekends(context: object, enabled: str) -> None:
    """Set standup skip_weekends in project configuration."""
    payload = _load_config_payload(context)
    payload.setdefault("standup", {})
    payload["standup"]["skip_weekends"] = enabled.lower() == "true"
    _write_config_payload(context, payload)


@given('standup timezone is "{timezone_name}"')
def given_standup_timezone(context: object, timezone_name: str) -> None:
    """Set standup timezone in project configuration."""
    context.standup_timezone_name = timezone_name
    payload = _load_config_payload(context)
    payload.setdefault("standup", {})
    payload["standup"]["timezone"] = timezone_name
    _write_config_payload(context, payload)


@given("the report time is fixed")
def given_report_time_is_fixed(context: object) -> None:
    """Pin standup report time for deterministic bucket tests."""
    _set_report_time(context, datetime(2026, 3, 10, 15, 0, tzinfo=timezone.utc))


@given("the report time is {weekday} {hour:d}:00 in standup timezone")
def given_report_time_in_standup_timezone(
    context: object, weekday: str, hour: int
) -> None:
    """Pin report time to a weekday and hour in standup timezone."""
    weekday_map = {
        "Monday": 0,
        "Tuesday": 1,
        "Wednesday": 2,
        "Thursday": 3,
        "Friday": 4,
        "Saturday": 5,
        "Sunday": 6,
    }
    target_weekday = weekday_map[weekday]
    timezone_info = _standup_timezone(context)
    base = datetime(2026, 3, 9, hour, 0, tzinfo=timezone_info)
    while base.weekday() != target_weekday:
        base += timedelta(days=1)
    _set_report_time(context, base.astimezone(timezone.utc))


@given('issue "{identifier}" has closed_at {hours:d} hours before report time')
def given_issue_closed_at_hours_before_report(
    context: object, identifier: str, hours: int
) -> None:
    """Set issue closed_at relative to the pinned report time."""
    project_dir = load_project_directory(context)
    issue = read_issue_file(project_dir, identifier)
    closed_at = _report_time(context) - timedelta(hours=hours)
    issue = issue.model_copy(update={"closed_at": closed_at, "status": "closed"})
    write_issue_file(project_dir, issue)


@given('issue "{identifier}" closed on the previous calendar day in standup timezone')
def given_issue_closed_previous_calendar_day(context: object, identifier: str) -> None:
    """Set issue closed_at on the previous calendar day in standup timezone."""
    project_dir = load_project_directory(context)
    issue = read_issue_file(project_dir, identifier)
    timezone_info = _standup_timezone(context)
    report_local = _report_time(context).astimezone(timezone_info)
    previous_day = report_local.date() - timedelta(days=1)
    closed_at = datetime(
        previous_day.year,
        previous_day.month,
        previous_day.day,
        16,
        0,
        tzinfo=timezone_info,
    ).astimezone(timezone.utc)
    issue = issue.model_copy(update={"closed_at": closed_at, "status": "closed"})
    write_issue_file(project_dir, issue)


@given(
    'issue "{identifier}" closed two calendar days before report day in standup timezone'
)
def given_issue_closed_two_calendar_days_before(
    context: object, identifier: str
) -> None:
    """Set issue closed_at two calendar days before the report day."""
    project_dir = load_project_directory(context)
    issue = read_issue_file(project_dir, identifier)
    timezone_info = _standup_timezone(context)
    report_local = _report_time(context).astimezone(timezone_info)
    closed_day = report_local.date() - timedelta(days=2)
    closed_at = datetime(
        closed_day.year,
        closed_day.month,
        closed_day.day,
        16,
        0,
        tzinfo=timezone_info,
    ).astimezone(timezone.utc)
    issue = issue.model_copy(update={"closed_at": closed_at, "status": "closed"})
    write_issue_file(project_dir, issue)


@given(
    'issue "{identifier}" closed on {weekday} before this Monday in standup timezone'
)
def given_issue_closed_on_weekday_before_monday(
    context: object, identifier: str, weekday: str
) -> None:
    """Set issue closed_at on a weekday before the report Monday."""
    weekday_map = {
        "Friday": 4,
        "Saturday": 5,
        "Sunday": 6,
        "Thursday": 3,
    }
    project_dir = load_project_directory(context)
    issue = read_issue_file(project_dir, identifier)
    timezone_info = _standup_timezone(context)
    report_local = _report_time(context).astimezone(timezone_info)
    days_back = (report_local.weekday() - weekday_map[weekday]) % 7
    if days_back == 0:
        days_back = 7
    closed_day = report_local.date() - timedelta(days=days_back)
    closed_at = datetime(
        closed_day.year,
        closed_day.month,
        closed_day.day,
        16,
        0,
        tzinfo=timezone_info,
    ).astimezone(timezone.utc)
    issue = issue.model_copy(update={"closed_at": closed_at, "status": "closed"})
    write_issue_file(project_dir, issue)


@given('the Kanbus configuration sets standup window to "{window}"')
def given_config_standup_window(context: object, window: str) -> None:
    """Set standup window in Kanbus configuration."""
    given_standup_window(context, window)


@given('the Kanbus configuration sets standup lookback to "{lookback}"')
def given_config_standup_lookback(context: object, lookback: str) -> None:
    """Set standup lookback in Kanbus configuration."""
    given_standup_lookback(context, lookback)


@given("the Kanbus configuration sets standup skip_weekends to {enabled}")
def given_config_standup_skip_weekends(context: object, enabled: str) -> None:
    """Set standup skip_weekends in Kanbus configuration."""
    given_standup_skip_weekends(context, enabled)


@when("I inspect standup window configuration")
def when_inspect_standup_window_configuration(context: object) -> None:
    """Resolve global standup window settings for assertions."""
    context.standup_window_settings = _resolved_window_settings(
        context,
        apply_profile_defaults=False,
    )


@when('I resolve standup window settings for profile "{profile}"')
def when_resolve_standup_window_settings(context: object, profile: str) -> None:
    """Resolve standup window settings for a profile."""
    context.standup_profile = profile
    context.standup_window_settings = _resolved_window_settings(context, profile)


@when("I resolve standup lookback duration")
def when_resolve_standup_lookback_duration(context: object) -> None:
    """Resolve lookback hours from configured lookback string."""
    payload = _load_config_payload(context)
    lookback = payload.get("standup", {}).get("lookback", "24h")
    context.standup_lookback_hours = parse_standup_lookback_hours(lookback)


@when("I resolve standup window settings in both runtimes")
def when_resolve_standup_window_settings_both_runtimes(context: object) -> None:
    """Resolve standup window settings in Python and Rust."""
    root = Path(context.working_directory)
    configuration = load_standup_configuration(root)
    overrides = getattr(context, "standup_window_overrides", StandupWindowOverrides())
    python_settings = resolve_standup_window_settings(configuration, None, overrides)
    context.python_window_settings = python_settings

    probe_binary = (
        Path(__file__).resolve().parents[3]
        / "rust"
        / "target"
        / "release"
        / "standup_window_probe"
    )
    if not probe_binary.is_file():
        build_result = subprocess.run(
            ["cargo", "build", "--quiet", "--bin", "standup_window_probe"],
            cwd=probe_binary.parents[2],
            capture_output=True,
            text=True,
            check=False,
        )
        if build_result.returncode != 0:
            raise AssertionError(build_result.stderr or build_result.stdout)
    result = subprocess.run(
        [str(probe_binary), str(root)],
        capture_output=True,
        text=True,
        check=False,
    )
    if result.returncode != 0:
        raise AssertionError(result.stderr or result.stdout)
    context.rust_window_settings = json.loads(result.stdout)


@then('standup window should be "{window}"')
def then_standup_window_should_be(context: object, window: str) -> None:
    """Assert resolved standup window mode."""
    settings = context.standup_window_settings
    assert settings.window == window


@then('standup lookback should be "{lookback}"')
def then_standup_lookback_should_be(context: object, lookback: str) -> None:
    """Assert resolved standup lookback string."""
    settings = context.standup_window_settings
    assert settings.lookback == lookback


@then("standup skip_weekends should be {enabled}")
def then_standup_skip_weekends_should_be(context: object, enabled: str) -> None:
    """Assert resolved standup skip_weekends flag."""
    expected = enabled.lower() == "true"
    settings = context.standup_window_settings
    assert settings.skip_weekends is expected


@then("standup lookback hours should be {hours:d}")
def then_standup_lookback_hours_should_be(context: object, hours: int) -> None:
    """Assert resolved lookback hours."""
    assert context.standup_lookback_hours == hours


@then('Python and Rust should agree on window "{window}"')
def then_python_rust_agree_on_window(context: object, window: str) -> None:
    """Assert Python and Rust agree on window mode."""
    assert context.python_window_settings.window == window
    assert context.rust_window_settings["window"] == window


@then("Python and Rust should agree on lookback hours {hours:d}")
def then_python_rust_agree_on_lookback_hours(context: object, hours: int) -> None:
    """Assert Python and Rust agree on lookback hours."""
    assert context.python_window_settings.lookback_hours == hours
    assert context.rust_window_settings["lookback_hours"] == hours


@then('the response should accept fields "{fields}"')
def then_response_accepts_fields(context: object, fields: str) -> None:
    """Assert console API accepts standup window request fields."""
    from kanbus.console_standup import StandupGenerateRequest

    payload = getattr(context, "last_post_json", {})
    accepted = StandupGenerateRequest.model_validate(payload)
    for field_name in (item.strip() for item in fields.split(",")):
        normalized = field_name.replace("and ", "").strip().strip('"')
        assert hasattr(accepted, normalized), f"missing accepted field: {normalized}"

    port = getattr(context, "console_server_port", None)
    if port is None:
        return
    url = f"http://127.0.0.1:{port}/api/standup"
    request = urllib.request.Request(
        url,
        data=json.dumps(payload).encode("utf-8"),
        headers={"Content-Type": "application/json"},
        method="POST",
    )
    try:
        with urllib.request.urlopen(request, timeout=10) as response:
            assert response.status == 200
    except urllib.error.HTTPError as error:
        body = error.read().decode("utf-8")
        if "unknown field" in body.lower() or "unexpected field" in body.lower():
            raise AssertionError(
                f"standup API rejected fields: {error.code} {body}"
            ) from error


@when('I POST "{path}" with JSON:')
def when_post_with_json(context: object, path: str) -> None:
    """Store JSON payload for a console API POST request."""
    context.last_post_path = path
    context.last_post_json = json.loads(context.text)


@then('command help should mention "{flag}"')
def then_command_help_should_mention(context: object, flag: str) -> None:
    """Assert CLI help mentions a flag."""
    stdout = context.result.stdout
    assert flag in stdout

File: rust/features/steps/standup_window_steps.rs
use std::collections::HashMap;
use std::fs;
use std::path::{Path, PathBuf};
use std::process::Command;

use chrono::{Datelike, Duration, TimeZone, Utc, Weekday};
use chrono_tz::Tz;
use cucumber::gherkin::Step;
use cucumber::{given, then, when};
use serde_json::Value;
use serde_yaml::{Mapping, Value as YamlValue};

use kanbus::config::default_project_configuration;
use kanbus::console_standup::StandupGenerateRequest;
use kanbus::models::IssueData;
use kanbus::standup::{load_standup_configuration, resolve_standup_profile};
use kanbus::standup_window::{
    parse_standup_lookback_hours, resolve_standup_report_time, resolve_standup_timezone,
    resolve_standup_window_settings, StandupWindowOverrides, StandupWindowSettings,
    STANDUP_REPORT_TIME_ENV,
};

use crate::step_definitions::initialization_steps::KanbusWorld;
use crate::step_definitions::query_steps::resolve_issue_project_directory;

fn repository_root(world: &KanbusWorld) -> PathBuf {
    world.working_directory.as_ref().expect("cwd").to_path_buf()
}

fn load_config_mapping(world: &KanbusWorld) -> Mapping {
    let config_path = repository_root(world).join(".kanbus.yml");
    let contents = fs::read_to_string(&config_path).expect("read config");
    serde_yaml::from_str(&contents).unwrap_or_else(|_| {
        let defaults = default_project_configuration();
        serde_yaml::to_value(defaults)
            .expect("serialize defaults")
            .as_mapping()
            .cloned()
            .expect("defaults mapping")
    })
}

fn write_config_mapping(world: &KanbusWorld, mapping: Mapping) {
    let config_path = repository_root(world).join(".kanbus.yml");
    let yaml = serde_yaml::to_string(&mapping).expect("serialize config");
    fs::write(config_path, yaml).expect("write config");
}

fn upsert_standup_field(world: &KanbusWorld, key: &str, value: YamlValue) {
    let mut mapping = load_config_mapping(world);
    let mut standup_block = mapping
        .get(&YamlValue::String("standup".to_string()))
        .and_then(YamlValue::as_mapping)
        .cloned()
        .unwrap_or_else(|| {
            let defaults = default_project_configuration();
            serde_yaml::to_value(defaults.standup)
                .expect("serialize standup defaults")
                .as_mapping()
                .cloned()
                .expect("standup mapping")
        });
    standup_block.insert(YamlValue::String(key.to_string()), value);
    mapping.insert(
        YamlValue::String("standup".to_string()),
        YamlValue::Mapping(standup_block),
    );
    write_config_mapping(world, mapping);
}

fn clear_report_time() {
    std::env::remove_var(STANDUP_REPORT_TIME_ENV);
}

fn set_report_time(report_time: chrono::DateTime<Utc>) {
    let text = report_time.to_rfc3339_opts(chrono::SecondsFormat::Secs, true);
    std::env::set_var(STANDUP_REPORT_TIME_ENV, text);
}

fn ensure_live_report_time() {
    if std::env::var(STANDUP_REPORT_TIME_ENV).is_err() {
        clear_report_time();
    }
}

fn standup_timezone(world: &KanbusWorld) -> Tz {
    if let Some(name) = world.standup_timezone_name.as_deref() {
        return name.parse::<Tz>().expect("timezone");
    }
    let configuration =
        load_standup_configuration(&repository_root(world)).expect("load standup config");
    resolve_standup_timezone(&configuration)
}

fn settings_to_value(settings: &StandupWindowSettings) -> Value {
    serde_json::json!({
        "window": settings.window,
        "lookback": settings.lookback,
        "lookback_hours": settings.lookback_hours,
        "skip_weekends": settings.skip_weekends,
        "timezone": settings.timezone.name(),
    })
}

fn resolved_window_settings(
    world: &KanbusWorld,
    profile: Option<&str>,
    apply_profile_defaults: bool,
) -> StandupWindowSettings {
    let root = repository_root(world);
    let configuration = load_standup_configuration(&root).expect("load standup config");
    let resolved_profile = if apply_profile_defaults {
        Some(resolve_standup_profile(profile).expect("resolve profile"))
    } else {
        None
    };
    resolve_standup_window_settings(
        &configuration,
        resolved_profile.as_deref(),
        &StandupWindowOverrides::default(),
    )
    .expect("resolve window settings")
}

fn read_issue_file(project_dir: &Path, identifier: &str) -> IssueData {
    let issue_path = project_dir
        .join("issues")
        .join(format!("{identifier}.json"));
    let contents = fs::read_to_string(&issue_path).expect("read issue");
    serde_json::from_str(&contents).expect("parse issue")
}

fn write_issue_file(project_dir: &Path, issue: &IssueData) {
    let issue_path = project_dir
        .join("issues")
        .join(format!("{}.json", issue.identifier));
    let contents = serde_json::to_string_pretty(issue).expect("serialize issue");
    fs::write(issue_path, contents).expect("write issue");
}

fn weekday_from_name(name: &str) -> Weekday {
    match name {
        "Monday" => Weekday::Mon,
        "Tuesday" => Weekday::Tue,
        "Wednesday" => Weekday::Wed,
        "Thursday" => Weekday::Thu,
        "Friday" => Weekday::Fri,
        "Saturday" => Weekday::Sat,
        "Sunday" => Weekday::Sun,
        other => panic!("unknown weekday: {other}"),
    }
}

#[given(expr = "standup window is {string}")]
fn given_standup_window(world: &mut KanbusWorld, window: String) {
    ensure_live_report_time();
    upsert_standup_field(world, "window", YamlValue::String(window));
}

#[given(expr = "standup lookback is {string}")]
fn given_standup_lookback(world: &mut KanbusWorld, lookback: String) {
    upsert_standup_field(world, "lookback", YamlValue::String(lookback));
}

#[given(expr = "standup skip_weekends is {word}")]
fn given_standup_skip_weekends(world: &mut KanbusWorld, enabled: String) {
    let value = enabled.eq_ignore_ascii_case("true");
    upsert_standup_field(world, "skip_weekends", YamlValue::Bool(value));
}

#[given(expr = "standup timezone is {string}")]
fn given_standup_timezone(world: &mut KanbusWorld, timezone_name: String) {
    world.standup_timezone_name = Some(timezone_name.clone());
    upsert_standup_field(world, "timezone", YamlValue::String(timezone_name));
}

#[given("the report time is fixed")]
fn given_report_time_is_fixed(_world: &mut KanbusWorld) {
    set_report_time(Utc.with_ymd_and_hms(2026, 3, 10, 15, 0, 0).unwrap());
}

#[given(expr = "the report time is {word} {int}:00 in standup timezone")]
fn given_report_time_in_standup_timezone(world: &mut KanbusWorld, weekday: String, hour: i64) {
    let target_weekday = weekday_from_name(&weekday);
    let timezone = standup_timezone(world);
    let mut base = timezone
        .with_ymd_and_hms(2026, 3, 9, hour as u32, 0, 0)
        .single()
        .expect("base report time");
    while base.weekday() != target_weekday {
        base += Duration::days(1);
    }
    set_report_time(base.with_timezone(&Utc));
}

#[given(expr = "issue {string} has closed_at {int} hours before report time")]
fn given_issue_closed_at_hours_before_report(
    world: &mut KanbusWorld,
    identifier: String,
    hours: i64,
) {
    let project_dir = resolve_issue_project_directory(world, &identifier);
    let issue = read_issue_file(&project_dir, &identifier);
    let report_time = resolve_standup_report_time().expect("report time");
    let closed_at = report_time - Duration::hours(hours);
    let updated = IssueData {
        closed_at: Some(closed_at),
        status: "closed".to_string(),
        ..issue
    };
    write_issue_file(&project_dir, &updated);
}

#[given(expr = "issue {string} closed two calendar days before report day in standup timezone")]
fn given_issue_closed_two_calendar_days_before(world: &mut KanbusWorld, identifier: String) {
    let project_dir = resolve_issue_project_directory(world, &identifier);
    let issue = read_issue_file(&project_dir, &identifier);
    let timezone = standup_timezone(world);
    let report_time = resolve_standup_report_time().expect("report time");
    let report_local = report_time.with_timezone(&timezone);
    let closed_day = report_local.date_naive() - Duration::days(2);
    let closed_at = timezone
        .from_local_datetime(&closed_day.and_hms_opt(16, 0, 0).expect("time"))
        .single()
        .map(|value| value.with_timezone(&Utc))
        .expect("closed_at");
    let updated = IssueData {
        closed_at: Some(closed_at),
        status: "closed".to_string(),
        ..issue
    };
    write_issue_file(&project_dir, &updated);
}

#[given(expr = "issue {string} closed on {word} before this Monday in standup timezone")]
fn given_issue_closed_on_weekday_before_monday(
    world: &mut KanbusWorld,
    identifier: String,
    weekday: String,
) {
    let target = weekday_from_name(&weekday);
    let project_dir = resolve_issue_project_directory(world, &identifier);
    let issue = read_issue_file(&project_dir, &identifier);
    let timezone = standup_timezone(world);
    let report_time = resolve_standup_report_time().expect("report time");
    let report_local = report_time.with_timezone(&timezone);
    let mut days_back = (report_local.weekday().num_days_from_monday() as i64
        - target.num_days_from_monday() as i64)
        .rem_euclid(7);
    if days_back == 0 {
        days_back = 7;
    }
    let closed_day = report_local.date_naive() - Duration::days(days_back);
    let closed_at = timezone
        .from_local_datetime(&closed_day.and_hms_opt(16, 0, 0).expect("time"))
        .single()
        .map(|value| value.with_timezone(&Utc))
        .expect("closed_at");
    let updated = IssueData {
        closed_at: Some(closed_at),
        status: "closed".to_string(),
        ..issue
    };
    write_issue_file(&project_dir, &updated);
}

#[given(expr = "the Kanbus configuration sets standup window to {string}")]
fn given_config_standup_window(world: &mut KanbusWorld, window: String) {
    given_standup_window(world, window);
}

#[given(expr = "the Kanbus configuration sets standup lookback to {string}")]
fn given_config_standup_lookback(world: &mut KanbusWorld, lookback: String) {
    given_standup_lookback(world, lookback);
}

#[given(expr = "the Kanbus configuration sets standup skip_weekends to {word}")]
fn given_config_standup_skip_weekends(world: &mut KanbusWorld, enabled: String) {
    given_standup_skip_weekends(world, enabled);
}

#[when("I inspect standup window configuration")]
fn when_inspect_standup_window_configuration(world: &mut KanbusWorld) {
    let settings = resolved_window_settings(world, None, false);
    world.standup_window_settings = Some(settings_to_value(&settings));
}

#[when(expr = "I resolve standup window settings for profile {string}")]
fn when_resolve_standup_window_settings(world: &mut KanbusWorld, profile: String) {
    let settings = resolved_window_settings(world, Some(&profile), true);
    world.standup_window_settings = Some(settings_to_value(&settings));
}

#[when("I resolve standup lookback duration")]
fn when_resolve_standup_lookback_duration(world: &mut KanbusWorld) {
    let mapping = load_config_mapping(world);
    let lookback = mapping
        .get(&YamlValue::String("standup".to_string()))
        .and_then(YamlValue::as_mapping)
        .and_then(|standup| standup.get(&YamlValue::String("lookback".to_string())))
        .and_then(YamlValue::as_str)
        .unwrap_or("24h");
    world.resolved_standup_lookback_hours =
        Some(parse_standup_lookback_hours(lookback).expect("parse lookback"));
}

#[when("I resolve standup window settings in both runtimes")]
fn when_resolve_standup_window_settings_both_runtimes(world: &mut KanbusWorld) {
    let root = repository_root(world);
    let configuration = load_standup_configuration(&root).expect("load standup config");
    let rust_settings =
        resolve_standup_window_settings(&configuration, None, &StandupWindowOverrides::default())
            .expect("resolve rust settings");
    world.rust_window_settings = Some(settings_to_value(&rust_settings));

    let repo_root = PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("..");
    let probe = repo_root.join("tools").join("standup_window_probe.py");
    let python_src = repo_root.join("python").join("src");
    let output = Command::new("python3")
        .env("PYTHONPATH", python_src)
        .arg(&probe)
        .arg(&root)
        .output()
        .expect("run python probe");
    assert!(
        output.status.success(),
        "python probe failed: {}",
        String::from_utf8_lossy(&output.stderr)
    );
    let payload: Value = serde_json::from_slice(&output.stdout).expect("parse python probe json");
    world.python_window_settings = Some(payload);
}

#[then(expr = "standup window should be {string}")]
fn then_standup_window_should_be(world: &mut KanbusWorld, window: String) {
    let settings = world.standup_window_settings.as_ref().expect("settings");
    assert_eq!(settings["window"].as_str().expect("window"), window);
}

#[then(expr = "standup lookback should be {string}")]
fn then_standup_lookback_should_be(world: &mut KanbusWorld, lookback: String) {
    let settings = world.standup_window_settings.as_ref().expect("settings");
    assert_eq!(settings["lookback"].as_str().expect("lookback"), lookback);
}

#[then(expr = "standup skip_weekends should be {word}")]
fn then_standup_skip_weekends_should_be(world: &mut KanbusWorld, enabled: String) {
    let expected = enabled.eq_ignore_ascii_case("true");
    let settings = world.standup_window_settings.as_ref().expect("settings");
    assert_eq!(
        settings["skip_weekends"].as_bool().expect("skip_weekends"),
        expected
    );
}

#[then(expr = "standup lookback hours should be {int}")]
fn then_standup_lookback_hours_should_be(world: &mut KanbusWorld, hours: u32) {
    assert_eq!(
        world
            .resolved_standup_lookback_hours
            .expect("lookback hours"),
        hours
    );
}

#[then(expr = "Python and Rust should agree on window {string}")]
fn then_python_rust_agree_on_window(world: &mut KanbusWorld, window: String) {
    let python = world.python_window_settings.as_ref().expect("python");
    let rust = world.rust_window_settings.as_ref().expect("rust");
    assert_eq!(python["window"].as_str().expect("python window"), window);
    assert_eq!(rust["window"].as_str().expect("rust window"), window);
}

#[then(expr = "Python and Rust should agree on lookback hours {int}")]
fn then_python_rust_agree_on_lookback_hours(world: &mut KanbusWorld, hours: u32) {
    let python = world.python_window_settings.as_ref().expect("python");
    let rust = world.rust_window_settings.as_ref().expect("rust");
    assert_eq!(
        python["lookback_hours"].as_u64().expect("python hours") as u32,
        hours
    );
    assert_eq!(
        rust["lookback_hours"].as_u64().expect("rust hours") as u32,
        hours
    );
}

#[when(regex = r#"^I POST \"([^\"]+)\" with JSON:$"#)]
fn when_post_with_json(world: &mut KanbusWorld, path: String, step: &Step) {
    world.last_post_path = Some(path);
    let text = step.docstring().expect("json docstring");
    world.last_post_json = Some(serde_json::from_str(text).expect("parse post json"));
}

#[then(
    regex = r#"^the response should accept fields \"([^\"]+)\", \"([^\"]+)\", and \"([^\"]+)\"$"#
)]
fn then_response_accepts_fields(
    world: &mut KanbusWorld,
    field_a: String,
    field_b: String,
    field_c: String,
) {
    let payload = world.last_post_json.as_ref().expect("post json").clone();
    let accepted: StandupGenerateRequest =
        serde_json::from_value(payload.clone()).expect("validate standup request");
    let mut present = HashMap::new();
    present.insert("window", accepted.window.is_some());
    present.insert("lookback", accepted.lookback.is_some());
    present.insert("skip_weekends", accepted.skip_weekends.is_some());
    for field in [field_a, field_b, field_c] {
        assert!(
            present.get(field.as_str()).copied().unwrap_or(false),
            "missing accepted field: {field}"
        );
    }

    let Some(port) = world.console_port else {
        return;
    };
    let url = format!("http://127.0.0.1:{port}/api/standup");
    let response = std::thread::spawn(move || {
        reqwest::blocking::Client::new()
            .post(url)
            .json(&payload)
            .send()
    })
    .join()
    .expect("http thread")
    .expect("post standup");
    assert_eq!(response.status().as_u16(), 200);
}

#[then(expr = "command help should mention {string}")]
fn then_command_help_should_mention(world: &mut KanbusWorld, flag: String) {
    let stdout = world.stdout.as_ref().expect("stdout");
    assert!(stdout.contains(&flag), "help missing {flag}");
}
